### PR TITLE
Add iconv:convert/3 API

### PR DESCRIPTION
**Pull request by MonsieurV**
Add the following API:

```
Output = iconv:convert("utf-8", "ascii", "123")
```

It will allows to use the module interchangeably / as a drop-in replacement for [processone/iconv](https://github.com/processone/iconv) or [emqplus/erlang-iconv](https://github.com/emqplus/erlang-iconv)

May be used for `gen_smtp`: see Vagabond/gen_smtp/pull/132

I'm not sure about the interface for the error case. For now I throw an exception with the original error name, but this may not be good. Does it preserve the original stack trace?
See https://github.com/zotonic/eiconv/commit/bf139960c00b899b9016afb543e5d416a468fb50#diff-2a6b2eb6299af8980f5114ff721322acR35

Peace

### CODE REVIEW

**Comment by mworrell** [comment]
@mmzeeman can you have a look at this? This fork is used by gen_smtp and these changes are used there to make the iconv library user pluggable.

**Comment by ddeboer** [comment]
It’s also used by Zotonic, so we need to make some changes there, too.

File: src/iconv.erl [hunk 1]
```diff
@@ -3,27 +3,39 @@
...
+% @doc Convert input which is in FromEncoding to ToEncoding.
+%
+convert(FromEncoding, ToEncoding, Input) ->
+    case eiconv:convert(FromEncoding, ToEncoding, Input) of
+        {ok, Output} -> Output;
+        {error, Error} ->
+            % throw the original error.
+            % TODO Let the badmatch occur?
+            % But it is not easy to catch by the user.
+            erlang:throw(Error)
```
Comment (ddeboer) [inline src/iconv.erl:40]: What kind of errors are these? Faulty `Input` or (e.g.) some missing server configuration? In the first case, returning an `{error, …}` tuple (so the user can handle the error or retry the call with some other input) is more appropriate.

**Review by ddeboer — APPROVED** [review]

File: src/iconv.erl (hunk 1 above)
Comment (MonsieurV) [reply src/iconv.erl:40]: Yes, this is for faulty inputs. As it is used for text encoding and may be submitted text the user do not control (like emails parsed by gen_smtp), it should be easy to handle the error case.

Actually, this is the interface adopted by `eiconv:convert/3` : https://github.com/zotonic/eiconv/blob/cf2c4b1d2e18da64cc691867b75143697b88980e/src/eiconv.erl#L103

I also find this interface more appropriate.

However here I added this `iconv/convert:3` function as a wrapper to be compatible with other iconv libraries. These other libraries ([processone/iconv](https://github.com/processone/iconv), [emqplus/erlang-iconv/](https://github.com/emqplus/erlang-iconv/blob/master/src/iconv.erl#L92)) have chosen as interface to return directly the output text.
(See Vagabond/gen_smtp/pull/132 for more context.)

I'm not sure we can return an `{error, ...}` for the error case and return the decoded text directly for the nominal case.

Note that the current `eiconv:convert/3` interface remains unchanged. Addind a wrapper here seems easier than submitted changes in other libraries that will break their APIs.

Comment (arjan) [reply src/iconv.erl:40]: > I'm not sure we can return an {error, ...} for the error case and return the decoded text directly for the nominal case.

No that is indeed not idiomatic. Either use return tuples to signify the success, or use a direct return value + throw on error.

File: README.md [hunk 2]
```diff
@@ -1,4 +1,87 @@
...
+{overrides,
+ [{override, eiconv,
+   [
+	{plugins, [pc]},
+	{port_env, [{"darwin|freebsd|openbsd", "LDFLAGS", "$LDFLAGS -liconv"},
+				{"freebsd|openbsd", "CFLAGS", "$CFLAGS -I/usr/local/include"},
+				{"freebsd|openbsd", "LDFLAGS", "$LDFLAGS -L/usr/local/lib"}]},
+
+	{port_specs, [{"priv/eiconv_nif.so", ["c_src/*.c"]}]},
+	{artifacts, ["priv/eiconv_nif.so"]},
```
Comment (ddeboer) [inline README.md:31]: Is this [still necessary with rebar3](https://github.com/blt/port_compiler#use-with-existing-dependency)? In Zotonic we don’t have this (only in gen_smtp) and everything works fine.

Comment (MonsieurV) [reply README.md:31]: There is no other build process whatsoever that handle that part?

I've tested again on https://github.com/YtoTech/geo-sensors-gateway, without this port_compiler configuration the `priv.eiconv_nif.so` is not built.

I'm new to that so I'm not sure I do that well.

Comment (ddeboer) [reply README.md:31]: Other repositories seem to [do the same](https://github.com/davisp/jiffy/blob/master/rebar.config), so I guess you’re right!

**Comment by mworrell** [comment]
Shouldn't we add this config to `eiconv` itself?

@mmzeeman

**Comment by ddeboer** [comment]
`eiconv` already has the `convert/3` function, but I’m all for changing the return values for `eiconv:convert/3` as well.

**Comment by MonsieurV** [comment]
Thanks!

**Pull request merged**